src/librustc_codegen_ssa/coverageinfo/map.rs

```diff
@@ -1,28 +1,96 @@
 use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::sync::Lrc;
+use rustc_span::source_map::{BytePos, Pos, SourceFile, SourceMap};
+
 use std::collections::hash_map;
+use std::fmt;
 use std::slice;
 
 #[derive(Copy, Clone, Debug)]
+#[repr(C)]
 pub enum CounterOp {
-    Add,
+    // Note the order (and therefore the default values) is important. With the attribute
+    // `#[repr(C)]`, this enum matches the layout of the LLVM enum defined for the nested enum,
+    // `llvm::coverage::CounterExpression::ExprKind`, as shown in the following source snippet:
+    // https://github.com/rust-lang/llvm-project/blob/f208b70fbc4dee78067b3c5bd6cb92aa3ba58a1e/llvm/include/llvm/ProfileData/Coverage/CoverageMapping.h#L146
     Subtract,
+    Add,
 }
 
+#[derive(Debug)]
 pub enum CoverageKind {
     Counter,
     CounterExpression(u32, CounterOp, u32),
 }
 
+#[derive(Debug)]
 pub struct CoverageSpan {
     pub start_byte_pos: u32,
     pub end_byte_pos: u32,
 }
 
+#[derive(Debug)]
 pub struct CoverageRegion {
     pub kind: CoverageKind,
     pub coverage_span: CoverageSpan,
 }
 
+/// A source code region used with coverage information.
+#[derive(Debug)]
+pub struct CoverageLoc {
+    /// Information about the original source file.
+    pub file: Lrc<SourceFile>,
+    /// The (1-based) line number of the region start.
+    pub start_line: u32,
+    /// The (1-based) column number of the region start.
+    pub start_col: u32,
+    /// The (1-based) line number of the region end.
+    pub end_line: u32,
+    /// The (1-based) column number of the region end.
+    pub end_col: u32,
+}
+
+impl fmt::Display for CoverageLoc {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // Customize debug format, and repeat the file name, so generated location strings are
+        // "clickable" in many IDEs.
+        write!(
+            f,
+            "{}:{}:{} - {}:{}:{}",
+            (*self).file.name,
+            self.start_line,
+            self.start_col,
+            (*self).file.name,
+            self.end_line,
+            self.end_col
+        )
+    }
+}
+
+fn lookup_file_line_col(source_map: &SourceMap, byte_pos: BytePos) -> (Lrc<SourceFile>, u32, u32) {
+    let found = source_map
+        .lookup_line(byte_pos)
+        .expect("should find coverage region byte position in source");
+    let file = found.sf;
+    let line_pos = file.line_begin_pos(byte_pos);
+
+    // Use 1-based indexing.
+    let line = (found.line + 1) as u32;
+    let col = (byte_pos - line_pos).to_u32() + 1;
+
+    (file, line, col)
+}
+
+impl CoverageRegion {
+    pub fn coverage_loc(&self, source_map: &SourceMap) -> CoverageLoc {
+        let (file, start_line, start_col) =
+            lookup_file_line_col(source_map, BytePos::from_u32(self.coverage_span.start_byte_pos));
+        let (_, end_line, end_col) =
```

Comment:
Maybe `assert!()` or `debug_assert!()` that these files are the same?

Reply:
@wesleywiser Yes, thanks! So I added this check, and surprisingly saw some cases (during ad hoc testing) where the assert failed. I was getting different filenames for the start and end of the byte_pos range!

But I've also seen some coverage results that show source ranges inside comments (not rustdoc code comments), so something is off sometimes.

It's almost surprising that a lot of things work correctly. So I still need to do more investigation.

For now, to avoid crashing, I changed this function to return an Option. Coverage mappings that have this problem get `None` from `coverage_loc()`, and I just skip them.

Reply:
@richkadel I wonder if any of those are related to macros? If you look at the MIR for a simple hello world ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1a8d62403d534afa14f56cd999691963)), you can see that some of the spans come from the playground file (`src/lib.rs`) while many of the others come from the expansion of the `println!()` macro and point into the std library (`/rustc/{git commit}/src/libstd/macros.rs`).